src/combinatorics/simple_graph/prod.lean

```diff
@@ -0,0 +1,154 @@
+and `H` is the graph on the product of the vertices such that `x` and `y` are related iff they agree
+on one component and the other one is related via either `G` or `H`. For example, the box product of
+two edges is a square.
+
+## Main declarations
+
+* `simple_graph.box_prod`: The box product.
+
+## Notation
+
+* `G □ H`: The box product of `G` and `H`.
+
+## TODO
+
+Define all other graph products!
+-/
+
+variables {α β γ : Type*}
+
+namespace simple_graph
+variables {G : simple_graph α} {H : simple_graph β} {I : simple_graph γ} {a a₁ a₂ : α} {b b₁ b₂ : β}
+  {x y : α × β}
+
+/-- Box product of simple graphs. It relates `(a₁, b)` and `(a₂, b)` if `G` relates `a₁` and `a₂`,
+and `(a, b₁)` and `(a, b₂)` if `H` relates `b₁` and `b₂`. -/
+def box_prod (G : simple_graph α) (H : simple_graph β) : simple_graph (α × β) :=
+{ adj := λ x y, G.adj x.1 y.1 ∧ x.2 = y.2 ∨ H.adj x.2 y.2 ∧ x.1 = y.1,
+  symm := λ x y, by simp [and_comm, or_comm, eq_comm, adj_comm],
+  loopless := λ x, by simp }
+
+infix ` □ `:70 := box_prod
+
+lemma box_prod_adj : (G □ H).adj x y ↔ G.adj x.1 y.1 ∧ x.2 = y.2 ∨ H.adj x.2 y.2 ∧ x.1 = y.1 :=
+iff.rfl
+
+@[simp] lemma box_prod_adj_left : (G □ H).adj (a₁, b) (a₂, b) ↔ G.adj a₁ a₂ :=
+by rw [box_prod_adj, and_iff_left rfl, or_iff_left (λ h : H.adj b b ∧ _, h.1.ne rfl)]
+
+@[simp] lemma box_prod_adj_right : (G □ H).adj (a, b₁) (a, b₂) ↔ H.adj b₁ b₂ :=
+by rw [box_prod_adj, and_iff_left rfl, or_iff_right (λ h : G.adj a a ∧ _, h.1.ne rfl)]
+
+variables (G H I)
+
+/-- The box product is commutative up to isomorphism. `equiv.prod_comm` as a graph isomorphism. -/
+@[simps] def box_prod_comm : G □ H ≃g H □ G := ⟨equiv.prod_comm _ _, λ x y, or_comm _ _⟩
+
+/-- The box product is associative up to isomorphism. `equiv.prod_assoc` as a graph isomorphism. -/
+@[simps] def box_prod_assoc : (G □ H) □ I ≃g G □ (H □ I) :=
+⟨equiv.prod_assoc _ _ _, λ x y, by simp only [box_prod_adj, equiv.prod_assoc_apply,
+  or_and_distrib_right, or_assoc, prod.ext_iff, and_assoc, @and.comm (x.1.1 = _)]⟩
+
+namespace walk
+variables {G}
+
+/-- Turn a walk on `G` into a walk on `G □ H`. -/
+def box_prod_left (b : β) : Π {a₁ a₂ : α}, G.walk a₁ a₂ → (G □ H).walk (a₁, b) (a₂, b)
+| _ _ nil := nil
+| _ _ (cons h w) := w.box_prod_left.cons $ box_prod_adj_left.2 h
+
+variables (G) {H}
+
+/-- Turn a walk on `H` into a walk on `G □ H`. -/
+def box_prod_right (a : α) : Π {b₁ b₂ : β}, H.walk b₁ b₂ → (G □ H).walk (a, b₁) (a, b₂)
+| _ _ nil := nil
+| _ _ (cons h w) := w.box_prod_right.cons $ box_prod_adj_right.2 h
+
+variables {G} [decidable_eq α] [decidable_eq β] [decidable_rel G.adj] [decidable_rel H.adj]
+
+/-- Project a walk on `G □ H` to a walk on `G` by discarding the moves in the direction of `H`. -/
+def of_box_prod_left : Π {x y : α × β}, (G □ H).walk x y → G.walk x.1 y.1
+| _ _ nil := nil
+| x z (cons h w) := or.by_cases h (λ hG, w.of_box_prod_left.cons hG.1)
+    (λ hH, show G.walk x.1 z.1, by rw hH.2; exact w.of_box_prod_left)
```

Comment:
Let's not develop it for this PR, but it seems like there ought to be some general interface lurking here. All I can think of at the moment is that if you have a graph homomorphism `G ->g H` and a left inverse `H -> G` (as plain functions) then you can pull walks back in a similar way.